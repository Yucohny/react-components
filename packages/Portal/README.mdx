---
name: Portal
---

# @c4605/react-portal

import { useState, useRef } from 'react'
import { Playground, Props } from 'docz'
import { Portal } from './src/Portal'
import { DocumentElement } from '../DocumentElement'
import { useInputValue } from '../DocHelpers/useInputValue'

## Introduction

Strong enough [`React.createPortal`](https://reactjs.org/docs/react-dom.html#createportal) replacement. Make most `React.createPortal` use case more easier.

## Properties

<Props of={Portal} />

## Usage

<Playground>
  {() => {
    const portalContentRef = useRef()

    const [visible, setVisible] = useState(false)
    
    const [renderInline, setRenderInline] = useInputValue(false, 'checked')

    const [closeOnClick, setCloseOnClick] = useInputValue(false, 'checked')

    const [closeOnOutsideClick, setCloseOnOutsideClick] = useInputValue(false, 'checked')

    const clickClose = event => {
      if (!(event.target instanceof HTMLElement)) return
      if (!portalContentRef.current) return
      if (closeOnClick) {
        return portalContentRef.current.contains(event.target)
      } else if (closeOnOutsideClick) {
        return !portalContentRef.current.contains(event.target)
      }
      return
    }

    return (
      <div>
        <p>
          <label>
            renderInline
            <input
              type="checkbox"
              checked={renderInline}
              onChange={setRenderInline}
            />
          </label>
        </p>
        <p>
          <label>
            closeOnClick
            <input
              type="checkbox"
              checked={closeOnClick}
              onChange={setCloseOnClick}
            />
          </label>
        </p>
        <p>
          <label>
            closeOnOutsideClick
            <input
              type="checkbox"
              checked={closeOnOutsideClick}
              onChange={setCloseOnOutsideClick}
            />
          </label>
        </p>
        <p>
          <button type="button" onClick={() => setVisible(true)}>
            Show Portal
          </button>
        </p>
        <DocumentElement style={visible ? { overflow: 'hidden' } : {}} />
        <Portal
          parent={renderInline ? null : undefined}
          visible={visible}
          clickClose={clickClose}
          onVisibleChange={setVisible}
        >
          <div
            style={{
              position: 'fixed',
              top: 0,
              right: 0,
              bottom: 0,
              left: 0,
              background: '#00000050',
            }}
          >
            <div
              ref={portalContentRef}
              style={{
                position: 'absolute',
                top: '50%',
                left: '50%',
                transform: 'translate(-50%, -50%)',
                minWidth: '300px',
                minHeight: '300px',
                background: '#fff',
              }}
            >
              <p>Portal content</p>
              <p>
                <button type="button" onClick={() => setVisible(false)}>
                  Close Portal
                </button>
              </p>
            </div>
          </div>
        </Portal>
      </div>
    )
  }}
</Playground>
