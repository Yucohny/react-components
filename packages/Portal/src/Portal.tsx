import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { SimpleJSON } from '@c4605/ts-types'
import shallowEqual from 'shallowequal'

export interface PortalProps {
  /** Specify the parent element for portal, if `null`, render children inline */
  parent: null | HTMLElement | (() => HTMLElement)
  /** Children in portal */
  children?: React.ReactNode

  /** Specify portal class name */
  className: string
  /** Specify portal style */
  style: React.CSSProperties & SimpleJSON

  /** The visibility of portal */
  visible: boolean
  /** Will be call when the visibility needs to be changed */
  onVisibleChange: (visible: boolean, reason: { event?: Event }) => void

  /**
   * Hide portal when function return true
   * @deprecated Use `onVisibleChange` instead
   */
  clickClose: (event: MouseEvent) => void | boolean
}

export class Portal extends React.PureComponent<PortalProps> {
  static displayName = 'Portal'

  static defaultProps = {
    parent: () => document.body,
    className: '',
    style: {},
    visible: true,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onVisibleChange() {},
    clickClose(event: MouseEvent) {
      /* istanbul ignore next */
      return event.button && event.button !== 0
    },
  }

  private _portalEl = document.createElement('div')

  /**
   * @deprecated
   */
  portal = this._portalEl

  componentDidMount(): void {
    document.addEventListener('click', this._onClickDocument)
    this._updatePortalEl(null, this.props)
  }

  componentDidUpdate(prevProps: PortalProps): void {
    this._updatePortalEl(prevProps, this.props)
  }

  componentWillUnmount(): void {
    document.removeEventListener('click', this._onClickDocument)
    this._updatePortalEl(this.props, null)
  }

  render(): React.ReactNode {
    if (this.props.parent) {
      return ReactDOM.createPortal(this.props.children, this._portalEl)
    }

    return (
      <div
        className={this.props.className}
        style={{
          ...this._getDisplayStyle(this.props.visible),
          ...this.props.style,
        }}
      >
        {this.props.children}
      </div>
    )
  }

  private _updatePortalEl(
    prevProps: PortalProps | null,
    nextProps: PortalProps,
  ): void
  private _updatePortalEl(
    prevProps: PortalProps,
    nextProps: PortalProps | null,
  ): void
  private _updatePortalEl(
    prevProps: PortalProps | null,
    nextProps: PortalProps | null,
  ): void {
    const portalEl = this._portalEl

    if (!prevProps || !nextProps || prevProps.parent !== nextProps.parent) {
      this._operateParent(prevProps, parent => parent?.removeChild(portalEl))
      this._operateParent(nextProps, parent => parent?.appendChild(portalEl))
    }

    if (nextProps) {
      if (!prevProps || prevProps.className !== nextProps.className) {
        portalEl.className = nextProps.className.trim() || ''
      }

      const prevStyle = prevProps && prevProps.style
      const nextStyle = nextProps.style
      if (!shallowEqual(prevStyle, nextStyle)) {
        portalEl.style.cssText = ''
        Object.assign(
          portalEl.style,
          this._getDisplayStyle(nextProps.visible),
          nextStyle,
        )
        Object.keys(nextStyle)
          .filter(p => p.startsWith('--'))
          .forEach(p => {
            portalEl.style.setProperty(p, nextStyle[p])
          })
      } else {
        if (nextProps.visible) {
          portalEl.style.removeProperty('display')
        } else {
          portalEl.style.display = 'none'
        }
      }
    }
  }

  private _getDisplayStyle(
    visible: boolean,
  ): { display?: React.CSSProperties['display'] } {
    if (visible) {
      return {}
    } else {
      return { display: 'none' }
    }
  }

  private _operateParent(
    props: PortalProps | null,
    operator: (parent: HTMLElement | null) => void,
  ): void {
    if (!props) return
    if (typeof props.parent === 'function') {
      operator(props.parent())
    } else {
      operator(props.parent)
    }
  }

  private _onClickDocument = (event: MouseEvent): void => {
    /* istanbul ignore next */
    if (!this.props.visible) return
    if (this.props.clickClose(event)) {
      this.props.onVisibleChange(false, { event })
    }
  }
}
