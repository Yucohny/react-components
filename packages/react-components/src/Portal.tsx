import { equals } from 'ramda'
import { SimpleJSON } from '@c4605-toolkit/ts-types'

export interface PortalProps {
  /** Specify the parent element for portal */
  parent: HTMLElement
  /** Children in portal */
  children: React.ReactNode

  /** Specify portal class  */
  className: string
  /** Specify portal style  */
  style: React.CSSProperties & SimpleJSON

  /** The visibility of portal */
  visible: boolean
  /** Will be call when the visibility need to be changed */
  onVisibleChange: (visible: boolean) => void

  /** Hide portal when function return true */
  clickClose: (event: MouseEvent) => void | boolean
}

export class Portal extends React.PureComponent<PortalProps> {
  static defaultProps = {
    parent: document.body,
    className: '',
    style: {},
    visible: false,
    onVisibleChange: () => {},
    clickClose: () => {},
  }

  portal = document.createElement('div')

  componentDidMount() {
    this.updateNode(null, this.props)
  }

  componentDidUpdate(prevProps: PortalProps) {
    this.updateNode(prevProps, this.props)
  }

  componentWillUnmount() {
    this.updateNode(this.props, null)
  }

  render() {
    return ReactDOM.createPortal(this.props.children, this.portal)
  }

  private updateNode(
    prevProps: PortalProps | null,
    nextProps: PortalProps,
  ): void
  private updateNode(
    prevProps: PortalProps,
    nextProps: PortalProps | null,
  ): void
  private updateNode(
    prevProps: PortalProps | null,
    nextProps: PortalProps | null,
  ) {
    const { portal } = this

    if (!prevProps || !nextProps || prevProps.parent !== nextProps.parent) {
      prevProps && prevProps.parent.removeChild(this.portal)
      nextProps && nextProps.parent.appendChild(this.portal)
    }

    document.removeEventListener('click', this.onClickDocument)
    if (
      (!prevProps && nextProps && nextProps.visible) ||
      (prevProps &&
        nextProps &&
        prevProps.visible !== nextProps.visible &&
        nextProps.visible)
    ) {
      document.addEventListener('click', this.onClickDocument)
    }

    if (nextProps) {
      if (!prevProps || prevProps.className !== nextProps.className) {
        portal.setAttribute('class', nextProps.className.trim() || '')
      }

      const prevStyle = prevProps && prevProps.style
      const nextStyle = nextProps.style
      if (!equals(prevStyle, nextStyle)) {
        portal.removeAttribute('style')
        Object.assign(
          portal.style,
          { display: nextProps.visible ? 'block' : 'none' },
          nextStyle,
        )
      } else {
        portal.style.display = nextProps.visible ? 'block' : 'none'
      }
    }
  }

  private onClickDocument = (event: MouseEvent) => {
    if (!this.props.visible) return
    const { portal } = this
    if (!portal || (event.button && event.button !== 0)) return
    if (this.props.clickClose(event)) {
      this.props.onVisibleChange(false)
    }
  }
}
