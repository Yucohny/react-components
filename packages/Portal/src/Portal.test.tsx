import * as React from 'react'
import { findDOMNode } from 'react-dom'
import { mount, ReactWrapper } from 'enzyme'
import { Portal, PortalProps } from './Portal'

describe('Portal', () => {
  let wrapper: ReactWrapper<PortalProps>

  afterEach(() => {
    try {
      wrapper.unmount()
    } catch {}
  })

  it('basicly works', () => {
    wrapper = mount(
      <div>
        <Portal className="portal" />
      </div>,
    )
    const portalEl = getPortalContainerEl(wrapper)
    expect(portalEl.parentElement).toBe(document.body)

    wrapper.unmount()
    expect(portalEl.parentElement).toBe(null)
    expect(document.querySelectorAll('.portal')).toHaveLength(0)
  })

  describe('support `parent` prop', () => {
    function assertParentPropWorks(
      wrapper: ReactWrapper<PortalProps>,
      parent: HTMLElement,
    ): void {
      const portalEl = getPortalContainerEl(wrapper)
      expect(portalEl.parentElement).toBe(parent)
      expect(findDOMNode(wrapper.instance())).toMatchSnapshot()
      wrapper.unmount()
      expect(portalEl.parentElement).toBe(null)
      expect(document.querySelectorAll('.portal')).toHaveLength(0)
    }

    it('as an element', () => {
      const parent = document.createElement('div')
      wrapper = mount(
        <div>
          <Portal parent={parent} />
        </div>,
      )
      assertParentPropWorks(wrapper, parent)
    })

    it('as a function', () => {
      const parent = document.createElement('div')
      wrapper = mount(
        <div>
          <Portal parent={() => parent} />
        </div>,
      )
      assertParentPropWorks(wrapper, parent)
    })

    it('as null', () => {
      wrapper = mount(
        <div>
          <Portal parent={null}>
            <div id="testEl" />
          </Portal>
        </div>,
      )
      expect(wrapper.find('#testEl')).toHaveLength(1)
      expect(findDOMNode(wrapper.instance())).toMatchSnapshot()
      wrapper.unmount()
      expect(wrapper.find('#testEl')).toHaveLength(0)
    })
  })

  function assertInAllRenderMode(
    asserts: (renderer: typeof render) => any,
  ): void {
    const parent = document.createElement('div')

    asserts(props => render({ ...props }))
    asserts(props => render({ ...props, parent }))
    asserts(props => render({ ...props, parent: () => parent }))
    asserts(props => render({ ...props, parent: null }))
  }

  it('support `className` prop', () => {
    assertInAllRenderMode(render => {
      const portalEl = getPortalContainerEl(
        render({ className: 'test-portal' }),
      )
      expect(portalEl.className).toBe('test-portal')
    })
  })

  it('support `style` prop', () => {
    assertInAllRenderMode(render => {
      const portalEl = getPortalContainerEl(
        render({
          style: {
            width: '10px',
            '--css-variable': '#eaeaea',
          },
        }),
      )
      expect(portalEl.style.cssText).toBe(
        'width: 10px; --css-variable: #eaeaea;',
      )
    })
  })

  it('support `visible` prop', () => {
    assertInAllRenderMode(render => {
      wrapper = render({ visible: false })
      const portalEl = getPortalContainerEl(wrapper)

      expect(portalEl.style.display).toBe('none')

      wrapper.setProps({ visible: true })
      expect(portalEl.style.display).toBe('')

      wrapper.setProps({ visible: false })
      expect(portalEl.style.display).toBe('none')
    })
  })

  it('support `clickClose` and `onVisibleChange` prop', () => {
    assertInAllRenderMode(render => {
      let clickCloseReturn: boolean | undefined
      const fakeClickEvent = new Event('click')
      const clickClose = jest.fn(() => clickCloseReturn)
      const onVisibleChange = jest.fn()
      wrapper = render({ visible: false, clickClose, onVisibleChange })
      const portalEl = getPortalContainerEl(wrapper)

      document.dispatchEvent(fakeClickEvent)
      expect(clickClose).not.toBeCalled()
      expect(portalEl.style.display).toBe('none')

      wrapper.setProps({ visible: true })

      document.dispatchEvent(fakeClickEvent)
      expect(clickClose).toHaveBeenCalledTimes(1)
      expect(clickClose).toHaveBeenCalledWith(fakeClickEvent)
      expect(onVisibleChange).not.toBeCalled()

      clickCloseReturn = false

      document.dispatchEvent(fakeClickEvent)
      expect(clickClose).toHaveBeenCalledTimes(2)
      expect(clickClose).toHaveBeenLastCalledWith(fakeClickEvent)
      expect(onVisibleChange).not.toBeCalled()

      clickCloseReturn = true

      document.dispatchEvent(fakeClickEvent)
      expect(clickClose).toHaveBeenCalledTimes(3)
      expect(clickClose).toHaveBeenLastCalledWith(fakeClickEvent)
      expect(onVisibleChange).toHaveBeenCalledTimes(1)
      expect(onVisibleChange).toHaveBeenCalledWith(
        false,
        expect.objectContaining({
          event: fakeClickEvent,
        }),
      )

      wrapper.unmount()
      document.dispatchEvent(fakeClickEvent)
      expect(clickClose).toHaveBeenCalledTimes(3)
      expect(onVisibleChange).toHaveBeenCalledTimes(1)
    })
  })
})

const getPortal = (wrapper: ReactWrapper<any>): Portal =>
  wrapper
    .find(Portal)
    .at(0)
    .instance() as Portal

const getPortalContainerEl = (wrapper: ReactWrapper<any>): HTMLDivElement => {
  const portal = getPortal(wrapper)
  if (portal.props.parent) {
    return portal['_portalEl']
  } else {
    return findDOMNode(portal) as HTMLDivElement
  }
}

const render = (props: Partial<PortalProps>): ReactWrapper<PortalProps> =>
  mount(<Portal {...props} />)
