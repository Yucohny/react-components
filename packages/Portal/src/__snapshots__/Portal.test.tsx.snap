// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal support \`parent\` prop as a function 1`] = `<div />`;

exports[`Portal support \`parent\` prop as an element 1`] = `<div />`;

exports[`Portal support \`parent\` prop as null 1`] = `
<div>
  <div
    class=""
  >
    <div
      id="testEl"
    />
  </div>
</div>
`;
